modules/auxiliary/gather/query_ldap.rb

```diff
@@ -0,0 +1,188 @@
+
+        when 'ENUM_EXCHANGE_SERVERS'
+          filter = Net::LDAP::Filter.construct('(&(objectClass=msExchExchangeServer)(!(objectClass=msExchExchangeServerPolicy)))') # Find Exchange Servers
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_EXCHANGE_RECIPIENTS'
+          # Find Exchange Recipients with or without fax addresses.
+          filter = Net::LDAP::Filter.construct('(|(mailNickname=*)(proxyAddresses=FAX:*))')
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_GROUPS'
+          # Standard LDAP groups query, followed by trying to find AD security groups, then trying to find Linux groups.
+          # Filters combined to remove duplicates.
+          filter = Net::LDAP::Filter.construct('(|(objectClass=group)(objectClass=groupOfNames)(groupType:1.2.840.113556.1.4.803:=2147483648)(objectClass=posixGroup))')
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_ORGUNITS'
+          filter = Net::LDAP::Filter.construct('(objectClass=organizationalUnit)') # Find OUs aka Organizational Units
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_ORGROLES'
+          filter = Net::LDAP::Filter.construct('(objectClass=organizationalRole)') # Find OUs aka Organizational Units
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_PEOPLE'
+          filter = Net::LDAP::Filter.construct('(objectClass=organizationalPerson)') # Find people within an organization by Person entries.
+          entries = perform_ldap_query(ldap, filter)
+
+        when 'ENUM_USERS'
+          # Common LDAP user query, followed by a query for AD User records by account type.
+          # Finally, query for Linux accounts by objectClass.
+          #
+          # Doing this all in one query also prevents duplicate entires across multiple queries.
+          filter = Net::LDAP::Filter.construct('(|(objectClass=inetOrgPerson)(objectClass=user)(sAMAccountType=805306368)(objectClass=posixAccount)(objectClass=GsAccount)(objectClass=GsSIPUser))')
+          entries = perform_ldap_query(ldap, filter)
+        end
+      end
+    rescue Rex::ConnectionTimeout => e
+      print_error("Could not query #{datastore['RHOST']}! Error was: #{e.message}")
+      return
+    end
+
+    return if entries[1].nil?
+
+    columns = []
+    entries[1].each do |entry|
+      entry.attribute_names.each do |attribute|
+        columns << attribute.to_s
+      end
+    end
+    columns.uniq!
+    tbl = Rex::Text::Table.new(
+      'Header' => "#{action.name} Dump of #{peer}",
+      'Indent' => 1,
+      'Columns' => columns
+    )
+    entries[1].each do |entry|
+      data = []
+      columns.each do |col|
+        col = col.to_sym
+        if entry[col].nil? || entry[col].empty? || entry[col][0].empty?
+          data << nil
+        else
+          data << entry[col].join(' || ')
+        end
+      end
+      tbl << data
+    end
+    print_status(tbl.to_s)
```

Comment:
I received a stack trace here. It looks like an encoding error.

```
msf6 auxiliary(gather/query_ldap) > show options 

Module options (auxiliary/gather/query_ldap):

   Name       Current Setting       Required  Description
   ----       ---------------       --------  -----------
   BASE_DN                          no        LDAP base DN if you already have it
   BIND_DN    aliddle@msflab.local  no        The username to authenticate to LDAP server
   BIND_PW    Password1             no        Password for the BIND_DN
   LDAPQUERY                        no        Query to run against the target LDAP server
   RHOSTS     192.168.159.10        yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
   RPORT      389                   yes       The target port
   SSL        false                 no        Enable SSL on the LDAP connection


Auxiliary action:

   Name                  Description
   ----                  -----------
   ENUM_ALL_OBJECTCLASS  Dump all objects containing any objectClass field.


msf6 auxiliary(gather/query_ldap) > run
[*] Running module against 192.168.159.10

[*] Discovering base DN automatically
[+] 192.168.159.10:389 Discovered base DN: DC=msflab,DC=local
[-] Auxiliary failed: ArgumentError invalid byte sequence in UTF-8
[-] Call stack:
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:413:in `split'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:413:in `block in chunk_values'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:410:in `each'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:410:in `each_with_index'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:410:in `each'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:410:in `map'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:410:in `chunk_values'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:395:in `row_to_s'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:125:in `block in to_s'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:121:in `each'
[-]   /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/rex-text-0.2.37/lib/rex/text/wrapped_table.rb:121:in `to_s'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/gather/query_ldap.rb:184:in `run'
[*] Auxiliary module execution completed
msf6 auxiliary(gather/query_ldap) > 
```

Reply:
Awesome, it looks like you've replicated https://github.com/rapid7/metasploit-framework/issues/15044 - @cgranleese-r7 looked into this a while ago and was never able to reproduce it.

If you still have the testing environment would you mind sharing the value.bytes and value.encoding that was causing the crash? I think that would be enough for a unit test. I'm guessing the implementation should be using `.chars` instead of `.split('')` at the very least

Reply:
Alright so at this point here https://github.com/rapid7/rex-text/blob/4d250b576d9a598ca79a05ea997175788ea81ccd/lib/rex/text/wrapped_table.rb#L411

I'm seeing that `value` appears to be a binary blob:

```
value.bytes = [133, 95, 156, 188, 16, 127, 17, 78, 142, 142, 235, 58, 84, 51, 65, 176]
value.encoding = #<Encoding:UTF-8>
```

FWIW, I was able to get around this by using `Rex::Text.to_hex_ascii`.

```diff
diff --git a/modules/auxiliary/gather/query_ldap.rb b/modules/auxiliary/gather/query_ldap.rb
index eec9f14544..4dd82e3851 100644
--- a/modules/auxiliary/gather/query_ldap.rb
+++ b/modules/auxiliary/gather/query_ldap.rb
@@ -175,9 +175,9 @@ class MetasploitModule < Msf::Auxiliary
         col = col.to_sym
         if entry[col].nil? || entry[col].empty? || entry[col][0].empty?
           data << nil
-        else
-          data << entry[col].join(' || ')
+          next
         end
+        data << Rex::Text.to_hex_ascii(entry[col].join(' || '))
       end
       tbl << data
     end
```

With that in place though, the output isn't exactly useful. I'm pretty sure it's due to the number of columns in the data.

![image](https://user-images.githubusercontent.com/58950994/175548793-262fc723-f81f-45a5-ba66-1b3581cdd30b.png)

Reply:
Thanks for finding a fix for this Spencer, might be good to use this as a temp fix until we can fix the underlying issue.

Reply:
Thanks for the byte dump @smcintyre-r7 :+1:
I've created a PR to fix the crash: https://github.com/rapid7/rex-text/pull/44

I haven't replicated this issue locally yet. But my latest thoughts are we might be dropping null bytes somewhere in the ldap logic - comment based on https://github.com/rapid7/metasploit-framework/issues/14955#issuecomment-1166177220

Reply:
Committed Spencer's fix already for this but going to try undo it and then test with @adfoster-r7's fix as per our Slack discussion so that we can see if https://github.com/rapid7/rex-text/pull/44 can be put into a Ready To Land state.

Reply:
With https://github.com/rapid7/rex-text/pull/44 now resolved n closed, I'm going to mark this issue as solved.